agent/api/task/task.go

```diff
@@ -2973,3 +3153,27 @@ func (task *Task) GetServiceConnectContainer() *apicontainer.Container {
 func (task *Task) IsServiceConnectEnabled() bool {
 	return task.GetServiceConnectContainer() != nil
 }
+
+// PopulateServiceConnectContainerMappingEnvVar populates APPNET_CONTAINER_MAPPING env var for AppNet Agent container
+// aka SC container
+func (task *Task) PopulateServiceConnectContainerMappingEnvVar() error {
```

Comment:
Do we have the logic for adding `LISTENER_PORT_MAPPING`? If not, that also needs to go here.

Reply:
We do, it's added at the end of ephemeral port selection logic [here](https://github.com/aws/amazon-ecs-agent/blob/feature/service-connect/agent/api/task/task.go#L578)

Reply:
Thanks. It would be good if we could move all the logic of adding Env variables into a single place.

Reply:
Hmm this may not be a straightforward change. 

So below events need to happen in the exact order
1. select ephemeral ports
2. create pause containers
3. populate `APPNET_CONTAINER_MAPPING`
4. create Appnet container 

`APPNET_LISTENER_PORT_MAPPING` is currently populated at step 1. While we technically can move it to step 3, it would be hard to tell which ports are originally specified in task def, vs. selected by ECS Agent (that's why we populate it the same time as we select the ephemeral ports).

Reply:
Could we not store this in a map? I would prefer to have all Env variables in a single place. But we can take this up after the first refactor.

Reply:
Agreed. We do store the newly assigned ephemeral ports as part of the container definition but we need to do mapping between the listener name from SC attachment and the containers to generate listener port mapping. I was suggesting that we store this mapping given we are processing the attachments in the first step. We could either do that or do the mapping again at 3rd step.

Reply:
The mapping already exist in `task.ServiceConnectConfig`. The env vars are just a transformation of this mapping. So this code looks good to me.

Reply:
Right now, we are adding 2 Env variables from Agent and we could add more in the future. So, it would be good if we could put those in a single place since these are the contracts/points of integration

Reply:
Lets ensure we revisit this in the refactor. TaskEngine / ServiceConnect should be a consolidation point. IF we want to add a Task / ServiceConnect package/file to consolidate on other things, that is fine. The overarching issue I see is that we have a LOT of SC specific code sprinkled in a must more generic place apitask.Task which is not a good separation of concerns.

Reply:
We are populating `APPNET_LISTENER_PORT_MAPPING` right after ephemeral port selection because after that we will lose the knowledge of which ports Agent actually selected, vs. pre-populated in task def. If we are to create `APPNET_LISTENER_PORT_MAPPING` the same time as `APPNET_CONTAINER_MAPPING`, we'll have to use another map for preserving the listener -> ephemeral port information. Unless, it's okay to put in the `APPNET_LISTENER_PORT_MAPPING` listeners with ephemeral ports and pre-defined ports, because then ECS Agent won't need to create the extra map for differentiating the listeners that got ephemeral ports.

Reply:
Yes. There should be no problem in passing in all the ports. We can handle that in EMS.

Reply:
@aws-gibbskt Yeah, have a TODO in this PR to move the function to SC manager
```
func (m *manager) initServiceConnectContainerMapping(task *apitask.Task, container *apicontainer.Container, hostConfig *dockercontainer.HostConfig) error {
	// TODO [SC] - Move the function here
	return task.PopulateServiceConnectContainerMappingEnvVar()
}
```
The refactoring PR I'm working on will extract away not only `PopulateServiceConnectContainerMappingEnvVar`, but also other SC-specific task handling from `task.go` and `docker_task_engine.go` which should contain only common logic. That PR should hopefully address your concern.

Reply:
@rajal-amzn Thanks for clarifying! Okay I'll move the two env var population into one place in the refactoring PR.